Add overview to README

nicholastmosher opened the pull request:
I started off by writing the post to the rust users forum, then I realized that what I was writing is relevant to anybody reading about flipper so I moved it to the README. I'd like some feedback, then I'll write a new post for the users forum which can then be more curated towards Flipper and Rust :)

georgemorgan commented on README.md:
```diff
@@ -2,28 +2,103 @@
+C/C++, Rust, Python, and Swift, and we plan on adding Objective-C, Java, and
+Javascript in the near future.
+
+# How it works
```
This is probably a bit too verbose for the top level README. This would be excellent to put in the Wiki though.

nicholastmosher replied:
On the contrary, I think Flipper is very difficult to explain, so by giving a quick walkthrough of all of the components, it helps to give a better idea of what exactly it does and why you would use it.

georgemorgan commented on README.md:
```diff
@@ -2,28 +2,103 @@
+hardware. All embedded platforms have limited resources, often orders of
+magnitude smaller than those available on a typical desktop computer. These
+limitations impose heavy constraints on the type of applications that you can
+write for these platforms. For example, nearly all embedded programs are
```
I had a hard time understanding the sentence starting with "For example". Also, all embedded devices to have access to heap memory, just less of It.

nicholastmosher replied:
Ah, yeah I agree. I'll go ahead and revise that

georgemorgan commented on README.md:
```diff
@@ -2,28 +2,103 @@
+
+# What's special about Flipper?
+
+Flipper has two core components: a lightweight operating system for the
```
The Flipper Message Runtime is just the protocol that conveys a remote procedure call.

nicholastmosher replied:
I know that, but I think that it's useful to assign nouns to concepts that need to be well-defined. It helps to anchor the concept in people's mind. I think that this is a good way to describe how the system works, even if it's not 100% true to the technical description that you and I think of. I think this paragraph almost moves in the direction of "Flipper is the physical board, FMR is the remote paradigm". By distinguishing these two I think it brings greater clarity to how it works.

I would personally like to keep this as it is, but if you have an alternate phrasing I'd like to hear it :)

georgemorgan commented on README.md:
```diff
@@ -2,28 +2,103 @@
+device, and a library to install on "host machines", such as the developer's
+computer. Together, these form a system that we like to call the Flipper
+Message Runtime, or FMR. FMR allows you to write programs that execute on the
+host machine but directly control the hardware. I like to think of it as
```
We should probably not have "I" in the top level README.

nicholastmosher replied:
Ah yeah, that's leftover from when it was going to be the users forum post. I'll take a look at it

georgemorgan commented on README.md:
```diff
@@ -2,28 +2,103 @@
+feature of Flipper is the ability to control it from any platform (Linux,
+MacOS, Windows, Android, or iOS), using any programming language.
+
+## An example
```
I think this would be better deferred to a readme in the examples folder.

nicholastmosher replied:
I think that people like to see how things work, and that examples and specific use cases draw them in. My favorite repositories to read through are the ones centered around "here's what YOU as the consumer of this code would do". I think that code examples in the top readme are not only a good idea, but very necessary to maintain people's attention and interest. You and I know what to expect when we talk about packages or language bindings, but nobody else will know what to picture in their minds. I don't even think this particular example is 100% accurate, but it's useful because it helps to build the understanding of how users will write code and how the code interacts in the overall system.

georgemorgan commented:
We should do some more iteration on this.

georgemorgan commented:
Any other comments on this?

nicholastmosher commented:
@georgemorgan you made a few small comments when I first opened this PR and I gave a thorough response to each one of your comments. So when you ask

> Any other comments on this?

I'm confused, because I've addressed all of the comments that you made but you haven't made any further arguments for making any changes. So I'm under the impression that this rendition of the README is pretty good, and I personally think it's good to merge unless you bring up specific things you'd like to see changed, or if you would like to write a draft I'd be happy to read that too.

georgemorgan commented:
Well I have voiced that I disagree that the README should be verbose and still stand strong on that.

DanGiaime commented:
I think the "readme_short" branch would be better for the top level readme

georgemorgan commented:
We need to get a “why” or “about us” page up on the Wiki first, but I think that indirection is the right approach here.

It would also be good to put pictures or videos in the Readme (as soon as we have them), while keeping the text to a minimum.

I will be working on a video this weekend hopefully.
